### PR TITLE
[BEAM-194] Add Java 8 examples archetype including mobile gaming example, with generator script

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @davorbonaci / @francesperry

**Comment by kennknowles** [comment]
Caveat: it is a bit of a copy/paste/modify job. The deps from the other archetype satisfied the gaming example, which means they need to be pruned.

**Comment by kennknowles** [comment]
Travis failures are all IWYU flake in starter archetype.

File: sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/DebuggingWordCount.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
+/*
```
Comment (davorbonaci) [inline sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/DebuggingWordCount.java]: This is the third copy of the wordcount pipelines; we should try to do something better.

File: sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/complete/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
+<!--
```
Comment (davorbonaci) [inline sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/complete/README.md]: I'd cut.

File: sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/complete/game/README.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
+<!--
```
Comment (davorbonaci) [inline sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/complete/game/README.md]: I'd cut.

File: sdks/java/maven-archetypes/java8-examples/pom.xml [hunk 4]
```diff
@@ -0,0 +1,94 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kennknowles) [inline sdks/java/maven-archetypes/java8-examples/pom.xml]: Done

File: sdks/java/maven-archetypes/java8-examples/pom.xml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.apache.beam</groupId>
+    <artifactId>beam-sdks-java-maven-archetypes-parent</artifactId>
+    <version>0.4.0-incubating-SNAPSHOT</version>
+    <relativePath>../pom.xml</relativePath>
+  </parent>
+
+  <artifactId>beam-sdks-java-maven-archetypes-java8-examples</artifactId>
```
Comment (kennknowles) [inline sdks/java/maven-archetypes/java8-examples/pom.xml]: done

File: sdks/java/maven-archetypes/java8-examples/pom.xml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  <name>Apache Beam :: SDKs :: Java :: Maven Archetypes :: Java 8 Examples</name>
```
Comment (kennknowles) [inline sdks/java/maven-archetypes/java8-examples/pom.xml]: Done

File: sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/pom.xml [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    <!-- Adds a dependency on a specific version of the Beam Google Cloud Platform IO module. -->
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>beam-sdks-java-io-google-cloud-platform</artifactId>
+      <version>0.4.0-incubating-SNAPSHOT</version>
+    </dependency>
+
+    <dependency>
+      <groupId>com.google.api-client</groupId>
+      <artifactId>google-api-client</artifactId>
```
Comment (kennknowles) [inline sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/pom.xml]: Confirmed that these are all dependencies of `examples-java8` or `examples-java`, which are both included in this archetype.

File: sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/DebuggingWordCount.java (hunk 1 above)
Comment (kennknowles) [inline sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/DebuggingWordCount.java]: Yes, it makes sense to not check these in. I don't know off the top of my head the best way to do that. Filed [BEAM-1005](https://issues.apache.org/jira/browse/BEAM-1005).

File: sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/complete/README.md (hunk 2 above)
Comment (kennknowles) [inline sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/complete/README.md]: Is it worth excluding the README?

File: sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/complete/game/README.md (hunk 3 above)
Comment (kennknowles) [inline sdks/java/maven-archetypes/java8-examples/src/main/resources/archetype-resources/src/main/java/complete/game/README.md]: Same

**Comment by kennknowles** [comment]
I have now also configured things so the content is generated, and removed sources from both archetypes. Unfortunately the way the configuration is inherited means that the `exec` plugin tries to execute again and fails, in some slightly curious way. One option is to move the execution to the parent, but that seems wrong. The other option is to wait until I have a few more minutes to dig into the behavior and see how to squash it.

File: .gitignore [hunk 8]
```diff
@@ -6,6 +6,10 @@
 target/
 bin/
 
+# Ignore generated archetypes
```
Comment (davorbonaci) [inline .gitignore]: I believe convention is that this goes to `${project.build.directory}/generated-sources/`, so it should not be necessary due to `target/` exclusion.

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml [hunk 9]
```diff
@@ -67,7 +67,24 @@
             <cleanupDaemonThreads>false</cleanupDaemonThreads>
           </configuration>
         </plugin>
-      </plugins>
+ 
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-dependency-plugin</artifactId>
```
Comment (davorbonaci) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml]: This doesn't belong here, right? Global plugin management?

File: sdks/java/maven-archetypes/pom.xml [hunk 10]
```diff
@@ -34,6 +34,7 @@
   <modules>
     <module>starter</module>
     <module>examples</module>
+    <module>examples-java8</module>
```
Comment (davorbonaci) [inline sdks/java/maven-archetypes/pom.xml]: Only on supported platforms?

**Comment by kennknowles** [comment]
Both archetypes are now autogenerated and I've had local success with verification. I have included a fix of some very old and presumably obsolete configuration that was breaking things. Jenkins had a transient failure but I have kicked it.

File: .gitignore (hunk 8 above)
Comment (kennknowles) [inline .gitignore]: I've spent a couple minutes looking for documentation on this, and I don't think [`archetype:create-from-project`](http://maven.apache.org/archetype/maven-archetype-plugin/create-from-project-mojo.html) has a way to support putting things in `generated-sources` to be instantiated to the user's `src` directory.

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml (hunk 9 above)
Comment (kennknowles) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml]: Done

File: sdks/java/maven-archetypes/pom.xml (hunk 10 above)
Comment (kennknowles) [inline sdks/java/maven-archetypes/pom.xml]: Done

**Comment by kennknowles** [comment]
It is worth noting that actually building the archetypes like this involves unix scripting so it not MS Windows friendly. We either need to check in the generated code or disable these processes on MS Windows.

**Comment by kennknowles** [comment]
I have updated the archetype parent to only run the unix-only builds on unix platforms. For now, at least this will make enthusiasts able to build the project on any platform, and users are not affected anyhow. If a Windows OS contributor would care to try to make this portable to Windows, that is welcome.

**Comment by davorbonaci** [comment]
This looks pretty awesome right now. High-level LGTM. Will take a closer look, but won't have anything more than low-level comments.

**Comment by davorbonaci** [comment]
LGTM

**Pull request closed without merging**